internal/cloud/instanceprotector/instance-protector.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/sirupsen/logrus"
+)
+
+type InstanceProtector struct {
+	protect   chan struct{}
+	unprotect chan struct{}
+}
+
+// NewInstanceProtector returns an instance protector that can
+// protect up to @size jobs in parallel. Beyond that limit
+// calls to Protect() starts to block.
+func NewInstanceProtector(size uint64) *InstanceProtector {
+	ip := &InstanceProtector{}
+	ip.protect = make(chan struct{}, size)
+	ip.unprotect = make(chan struct{}, size)
+	return ip
+}
+
+func (ip *InstanceProtector) Protect() {
+	ip.protect <- struct{}{}
+}
+
+func (ip *InstanceProtector) Unprotect() {
+	ip.unprotect <- struct{}{}
+}
+
+func (ip *InstanceProtector) Start(ctx context.Context) {
+	// Perform alternate calls to protect / unprotect until
+	// ctx is cancelled.
+	//
+	// Collate matching protect/unprotect tokens when possible
+	// to avoid uneccesary calls (which can be slow), this should
+	// minimize the risk of queues of tokens building up which in
+	// the worst case could block new work being scheduled, or at
+	// least cause us to unprotect when we should still be protected
+	// and vice versa.
+	//
+	// The main reason for doing this async, is that the calls can
+	// be very slow and cause unneccesary blocking. There is already
+	// a race between a job being dequeued and the instance being
+	// protected, and moreover the protection is not always respected
+	// by the cloud provider, so we do not lose much by being async.
+	for {
+	protect:
+		for {
+			// Pop off alternating protect / unprotect tokens until we receive
+			// a protect token without a matching unprotect one for ten seconds.
+			// At that point, protect the instance and break from this loop.
+			//
+			// We wait for ten seconds to avoid a busy loop, and allow more
+			// collation. At worst this means we have to redo ten seconds of
+			// work in case the instance is killed early (but that can anyway
+			// happen for other reasons).
+			select {
+			case <-ip.protect:
+			case <-ctx.Done():
+				return
+			}
+			select {
+			case <-ip.unprotect:
+			case <-time.After(10 * time.Second):
+				setProtection(true)
+				break protect
+			case <-ctx.Done():
+				return
+			}
+		}
+	unprotect:
+		for {
+			// Pop off alternating unprotect / protect tokens until we receive
+			// an unprotect token without a matching protect one for one minute.
+			// At that point, onprotect the instance and break from this loop and
+			// start waiting for a protect token again.
+			//
+			// We wait for one minute to avoid a busy loop, and allow more
+			// collation. At worst this means we keep an instance running for
+			// one more minute than we need to.
+			select {
+			case <-ip.unprotect:
+			case <-ctx.Done():
+				return
+			}
+			select {
+			case <-ip.protect:
+			case <-time.After(time.Minute):
+				setProtection(false)
+				break unprotect
+			case <-ctx.Done():
+				return
```

Comment:
I wonder if we should not call `setProtection(false)` here as well. If I'm reading the code correctly, it could happen that we protected the VM as part of starting a job, then finished the job, but if the `ctx` is canceled before the minute passes, we would keep the VM protected and return. 🤔  Given the comments, it does not sound like the cloud provider would care anyway.

Reply:
D'oh! Very good point! I think you are right.

Reply:
Fixed.